synapse/storage/e2e_room_keys.py

```diff
@@ -0,0 +1,240 @@
+                    "session_data": room_key['session_data'],
+                },
+                lock=False,
+            )
+
+            return True
+
+        return self.runInteraction(
+            "set_e2e_room_key", _set_e2e_room_key_txn
+        )
+
+    # XXX: this isn't currently used and isn't tested anywhere
+    # it could be used in future for bulk-uploading new versions of room_keys
+    # for a user or something though.
+    def set_e2e_room_keys(self, user_id, version, room_keys):
+
+        def _set_e2e_room_keys_txn(txn):
+
+            values = []
+            for room_id in room_keys['rooms']:
+                for session_id in room_keys['rooms'][room_id]['sessions']:
+                    session = room_keys['rooms'][room_id]['sessions'][session_id]
+                    values.append(
+                        {
+                            "user_id": user_id,
+                            "room_id": room_id,
+                            "session_id": session_id,
+                            "version": version,
+                            "first_message_index": session['first_message_index'],
+                            "forwarded_count": session['forwarded_count'],
+                            "is_verified": session['is_verified'],
+                            "session_data": session['session_data'],
+                        }
+                    )
+
+            self._simple_insert_many_txn(
+                txn,
+                table="e2e_room_keys",
+                values=values
+            )
+
+            return True
+
+        return self.runInteraction(
+            "set_e2e_room_keys", _set_e2e_room_keys_txn
+        )
+
+    @defer.inlineCallbacks
+    def get_e2e_room_keys(self, user_id, version, room_id, session_id):
+
+        keyvalues = {
+            "user_id": user_id,
+            "version": version,
+        }
+        if room_id:
+            keyvalues['room_id'] = room_id
+        if session_id:
+            keyvalues['session_id'] = session_id
+
+        rows = yield self._simple_select_list(
+            table="e2e_room_keys",
+            keyvalues=keyvalues,
+            retcols=(
+                "user_id",
+                "room_id",
+                "session_id",
+                "first_message_index",
+                "forwarded_count",
+                "is_verified",
+                "session_data",
+            ),
+            desc="get_e2e_room_keys",
+        )
+
+        # perlesque autovivification from https://stackoverflow.com/a/19829714/6764493
+        class AutoVivification(dict):
+            def __getitem__(self, item):
+                try:
+                    return dict.__getitem__(self, item)
+                except KeyError:
+                    value = self[item] = type(self)()
+                    return value
+
+        sessions = AutoVivification()
+        for row in rows:
+            sessions['rooms'][row['room_id']]['sessions'][row['session_id']] = {
```

Comment:
I think this can be written as:

```python
sessions['rooms'].setdefault(row['room_id'], {"sessions":{}})['sessions'][row['session_id']] = { ... }
```

Or if  you don't mind that we use an extra line:

```python
room_entry = sessions['rooms'].setdefault(row['room_id'], {"sessions":{}})
room_entry['sessions'][row['session_id']] = { ... }
```

Reply:
thanks; done